### DISCUSSION TITLE
How to configure Azure OpenAI

### DISCUSSION THREAD

**Discussion by tsoernes** [zed-industries/zed#20232] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I have tried this:
```json
  "language_models": {
    "openai": {
      "version": "1",
      "api_url": "https://XXXXXX.openai.azure.com/openai/deployments/m-gpto-06082024",
      "available_models": [
        {
          "api_version": "2024-08-01-preview",
          "_name": "m-gpto-06082024",
          "name": "m-gpt-4o-mini-18072024",
          "max_tokens": 128000
        }
      ]
    }
  },
  "assistant": {
    "default_model": {
      "provider": "openai",
      "model": "m-gpt-4o-mini-18072024"
    },
    "version": "2"
  }
```
But when selecting the model in the assistants sidebar it just says "LLM provider is not configured."  and I'm not able to interact with it.

How do I provide the API key for Azure?

### Environment

Zed: v0.159.10 (Zed)
OS: Linux Wayland fedora 40
Memory: 30.9 GiB
Architecture: x86_64
GPU: Intel(R) Graphics (RPL-U) || Intel open-source Mesa driver || Mesa 24.1.7

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by tsoernes** [comment]
If you need a free Azure endpoint for testing you can create a trial azure account: https://azure.microsoft.com/en-us/pricing/purchase-options/azure-account

**Comment by jonmagic** [comment]
I'm also interested in this and willing to work on the implementation but wanted to follow the contribution documentation which says to start with a discussion first so I found this discussion and also https://github.com/zed-industries/zed/discussions/25544.

Here's an explanation of the differences and what my next steps would be:

### 🔑 Authentication

* **OpenAI**: Uses a straightforward API key.
* **Azure OpenAI**: Supports both API keys and Microsoft Entra ID (formerly Azure Active Directory) tokens. Entra ID is recommended for enhanced security and enterprise integration.

### 🌐 Endpoint Configuration

* **OpenAI**: Endpoints are standardized, e.g., `https://api.openai.com/v1/...`.
* **Azure OpenAI**: Endpoints are specific to your Azure resource, e.g., `https://<your-resource-name>.openai.azure.com/openai/deployments/<deployment-id>/...`. You must specify the API version via the `api-version` query parameter.

### 🧠 Model Deployment

* **OpenAI**: Models are referenced directly by name, such as `"gpt-4"`.
* **Azure OpenAI**: Models must be deployed within your Azure resource. API calls reference the deployment name, not the base model name.

### ⚙️ Client Initialization

**OpenAI:**

```python
from openai import OpenAI
client = OpenAI(api_key="your-openai-api-key")
```

**Azure OpenAI:**

```python
from openai import AzureOpenAI
client = AzureOpenAI(
    api_key="your-azure-api-key",
    azure_endpoint="https://<your-resource-name>.openai.azure.com",
    api_version="2024-07-01-preview"
)
```

**For Entra ID authentication:**

> [!NOTE]
> I probably wouldn't implement this authentication strategy because I do not need it at this time.

```python
from azure.identity import DefaultAzureCredential, get_bearer_token_provider
from openai import AzureOpenAI

token_provider = get_bearer_token_provider(
    DefaultAzureCredential(), "https://cognitiveservices.azure.com/.default"
)

client = AzureOpenAI(
    azure_endpoint="https://<your-resource-name>.openai.azure.com",
    api_version="2024-07-01-preview",
    azure_ad_token_provider=token_provider
)
```

### 📦 Request Differences

* **Model Parameter**:

  * OpenAI: Use the `model` parameter with the model name.
  * Azure OpenAI: Use the `deployment_id` parameter with your deployment name.
* **API Versioning**: Azure requires specifying the `api-version` in the query parameters.

### 📊 Behavioral Nuances

* **Determinism**: Azure OpenAI's embeddings (e.g., `text-embedding-3-small`) are more deterministic, returning consistent outputs for identical inputs.
* **Output Variability**: Azure may produce shorter or more conservative completions due to stricter content filters.
* **Latency & Throughput**: Azure often offers lower latency and higher throughput, especially when using regional deployments.

### 🛠️ Implementation Strategy

To support both APIs:

1. **Abstract Configuration**: Create a configuration layer to switch between OpenAI and Azure OpenAI settings (API keys, endpoints, etc.).
2. **Dynamic Client Initialization**: Initialize the appropriate client based on the selected configuration.
3. **Parameter Handling**: Adjust request parameters to align with the selected API's requirements (e.g., `model` vs. `deployment_id`).
4. **Error Handling**: Implement robust error handling to manage differences in API responses and potential exceptions.
5. **Testing**: Thoroughly test both configurations to ensure consistent behavior across APIs.

**Reply by JairajSustained** [reply]
I also use ZED and have Azure AI Foundry access.
How can I help?

**Comment by techy-0** [comment]
Has any progress been made on this? I made a proxy that applies the API version as a URL parameter, logs into Azure using their library, and apply authentication since my instance of Azure OpenAI doesn't support API keys, we are required to use Microsoft Entra. This is a hacky fix and I keep getting an error saying `Error data did not match any variant of untagged enum ResponseStreamResult` which often results in truncated responses. Also, any idea what is causing this issue for me?

**Comment by tomprogramuje** [comment]
don't know if anyone still fumbles, but it actually can be done, trick is to not use the standard azure openai endpoint, but the one for openai sdk (you can find it in your model deployment):

E.g.

```
"language_models": {
   "openai_compatible": {
     "Azure OpenAI": {
       "api_url": "https://<your-deployment>.openai.azure.com/openai/v1/",
       "available_models": [
         {
           "name": "gpt-5",
           "display_name": null,
           "max_tokens": 200000,
           "max_output_tokens": 32000,
           "max_completion_tokens": 200000,
           "capabilities": {
             "tools": true,
             "images": false,
             "parallel_tool_calls": false,
             "prompt_cache_key": false
           }
         }
       ]
     }
    }
```

i made it all including api key via the UI that emitted the above settings but apparently it could be done directly in settings plus api key exposed as env variable as stated here: https://zed.dev/docs/ai/llm-providers#openai-api-compatible

**Reply by ZeroThe2nd** [reply]
I do, as my organisation uses the enterprise version, which has a _slightly different_ endpoint structure, and requires a custom header to be set with the API token. (which IMO is bad design, but I digress).

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links

**Comment by frhanjav** [comment]
you can follow this: https://learn.microsoft.com/en-us/azure/foundry/openai/how-to/codex?tabs=npm

and add this to the ~/.codex/config.toml

```
model = "gpt-5-codex"  # Replace with your actual Azure model deployment name
model_provider = "azure"
model_reasoning_effort = "medium"

[model_providers.azure]
name = "Azure OpenAI"
base_url = "https://YOUR_RESOURCE_NAME.openai.azure.com/openai/v1"
env_key = "AZURE_OPENAI_API_KEY"
wire_api = "responses"
```

and add this: `export AZURE_OPENAI_API_KEY="your_api_key_here"` to the shell config and you'll be able to use the default `Codex External CLI Agent` instead of having to use the Configure the LLM provider by adding a Compatible OpenAI provider (also saves from the trouble of having to add: "effort": "high", etc).
